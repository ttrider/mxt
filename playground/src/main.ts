import { observable, computed } from "mobx"

import { ex01 } from "./templates/ex01";
import { ex02 } from "./templates/ex02";
import { ex03 } from "./templates/ex03";

class TestData {
    @observable r = 200;
    @observable g = 18;
    @observable b = 228;

    @computed get color() {
        return "#" + this.r.toString(16) + this.g.toString(16) + this.b.toString(16);
    }

    colorClick() {

        this.r += 20;
        this.g += 15;
        this.b += 35;

        this.r = this.r % 255;
        this.g = this.g % 255;
        this.b = this.b % 255;

        if (this.el){
            //this.el.dispose();

            const root_mv = document.getElementById("mv");
            if (root_mv) {
                this.el.appendTo(root_mv);
            }
        }
    }

    el:any
}


const data = new TestData();

if (document) {

    // example 01
    const root_ex01 = document.getElementById("ex01");
    if (root_ex01) {

        data.el = ex01(data);
        data.el.appendTo(root_ex01);

    }

    // example 02
    const root_ex02 = document.getElementById("ex02");
    if (root_ex02) {
        ex02(data).appendTo(root_ex02);
    }

    // example 03
    const root_ex03 = document.getElementById("ex03");
    ex03(data, root_ex03);
}

