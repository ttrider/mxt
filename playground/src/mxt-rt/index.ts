export * from "./types";
export * from "./context";
export * from "./template-context";
export * from "./data-context";
export * from "./parts-context";
export * from "./loop-context";
export * from "./styles-context";