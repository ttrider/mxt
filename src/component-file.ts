import { ElementInfo, StyleElementInfo, TemplateInfo } from ".";
import { parseDOM } from "htmlparser2";
import * as dom from "domhandler";
import fs from "fs";
import path from "path";
import util from "util";
import ts from "typescript";
import * as d from "./ts";

const readFile = util.promisify(fs.readFile);


export class ComponentFile {

    filePath: string;
    content: string;
    dom: dom.Node[];
    links: ElementInfo[] = [];
    globalStyles: StyleElementInfo[] = [];

    templates: { [id: string]: TemplateInfo } = {};
    errors: Error[] = [];

    initStatements = d.StatementList();
    componentStatements = d.StatementList();

    importStatements: {
        [from: string]: { name: string, as: string }[]
    } = {};

    private constructor(content: string, filePath: string) {

        this.content = content;
        this.dom = parseDOM(this.content, { xmlMode: true, withStartIndices: true, withEndIndices: true });
        this.filePath = filePath;
    }

    addImport(identifier: string | { name: string, as: string } | Array<string | { name: string, as: string }>, from: string) {

        if (!Array.isArray(identifier)) {
            identifier = [identifier];
        }
        const definitions = identifier.map<{ name: string, as: string }>(i => {
            if (typeof i === "string") {
                i = { name: i, as: i };
            }
            return i;
        });


        const fromSet = this.importStatements[from];
        if (!fromSet) {
            this.importStatements[from] = definitions;
            return;
        }

        for (const definition of definitions) {
            // lookup existing definition
            const existing = fromSet.find((item) => item.as === definition.as);
            if (!existing) {
                fromSet.push(definition);
            } else {

                if (definition.name != existing.name) {
                    throw new Error(`redefinition of the import: '${existing.name}' by ${definition.name}`);
                }
            }
        }
        return this;
    }

    getImportStatements() {

        const items: ts.Statement[] = [];

        for (const from in this.importStatements) {
            if (this.importStatements.hasOwnProperty(from)) {

                const ii = d.ImportStatement(this.importStatements[from], from);

                items.push(ii);
            }
        }
        return items;
    }

    getStatements() {
        const s =
            [
                ...this.getImportStatements(),
                ...this.initStatements,
                ...this.componentStatements
            ] as ts.Statement[];
        return s;
    }






    static fromContent(content: string) {
        return new ComponentFile(content, "./dummy.html")
    }

    static async fromFile(filePath: string) {
        filePath = path.resolve(filePath);
        const fileBuffer = await readFile(filePath);
        if (!fileBuffer) {
            throw new Error("can't read the file: " + filePath);
        }
        const content = fileBuffer.toString();
        return new ComponentFile(content, filePath);
    }
}
