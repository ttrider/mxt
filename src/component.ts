import { TokenizedContent, TemplateInfo } from ".";
import { Part, PartReference } from "./template-parts";

export class Component {

    static partIndex: number = 0;
    static newPartId() {
        return `p${Component.partIndex++}`;
    }


    dynamicStyles?: TokenizedContent[];
    styles?: TokenizedContent[];
    id: string;
    rootPart?: PartReference;
    parts: { [id: string]: TemplateInfo } = {};

    importParts: { [id: string]: string } = {};

    partset: { [id: string]: Part } = {};

    constructor(id: string) {
        this.id = id;
    }

    addComponentImport(name: string) {
        if (this.importParts[name]) {
            return this.importParts[name];
        }
        return this.importParts[name] = Component.newPartId();
    }

    newPart<T>(init?: T, id?: string): Part & T {
        if (!id) {
            id = Component.newPartId();
        }

        const part = (init) ? {
            id,
            ...init
        } as Part & T : { id } as Part & T;

        this.partset[id] = part;

        return part;
    }

}