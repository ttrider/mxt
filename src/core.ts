import { ElementInfo, StyleElementInfo, TemplateInfo } from ".";
import { Statement } from "@babel/types";
import * as dom from "domhandler";


export interface ComponentFile {

    path: string,
    name: string,
    content: string,
    dom: dom.Node[],
    links: ElementInfo[],
    globalStyles: StyleElementInfo[],

    templates: { [id: string]: TemplateInfo },
    errors: Error[],

    initStatements: Statement[],
    componentStatements: Statement[],
}
