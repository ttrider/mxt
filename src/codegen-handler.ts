import { HandlerContext, AttributeTokenInfo } from ".";
import { returnStatement, unaryExpression, ifStatement, expressionStatement, identifier, callExpression } from "@babel/types";
import * as t from "@babel/types";
import { statementList, declareFunction, declareVar, declareObjectDestruction, makeTemplateLiteral, makeAssignment, makeThrow, makeCall } from "./ast/builder";
import { ComponentFile } from "./component-file";


export function codegen(context: HandlerContext, componentFile: ComponentFile) {

    componentFile.addImport("autorun", "mobx");

    // create definitions code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];

            const constTemplateName = `${template.id}$$template`;

            const templateLiteral = makeTemplateLiteral(template.elements);


            componentFile.initStatements.add(

                declareVar(constTemplateName)
                    .const
                    .init(makeCall("document.createElement", "template").build())
                    .build(),

                makeAssignment(constTemplateName + ".innerHTML", templateLiteral)
            );

        }
    }

    // create init function code
    for (const templateId in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(templateId)) {
            const template = componentFile.templates[templateId];

            const constTemplateName = `${template.id}$$template`;

            const funcBody = statementList().add(
                declareVar("component")
                    .const
                    .init(makeCall("document.importNode", t.identifier(constTemplateName), true).build())
                    .build()
            );

            const elements = template.tokens.reduce((p, v) => {
                if (p[v.elementId] === undefined) {
                    p[v.elementId] = [v];
                } else {
                    p[v.elementId].push(v);
                }
                return p;
            }, {} as { [elementId: string]: AttributeTokenInfo[] });


            for (const elementId in elements) {
                if (elements.hasOwnProperty(elementId)) {
                    const tokenSet = elements[elementId];
                    const elementOriginalId = tokenSet[0].elementIdOriginal;

                    const elementName = `${elementId}$$element`;

                    const externalReferences = Object.keys(tokenSet.reduce<{ [name: string]: any }>((e, i) => {
                        for (const er of i.externalReferences) {
                            e[er] = null;
                        }
                        return e;
                    }, {})).map(er => { return { name: er }; });





                    funcBody
                        .add(declareVar(elementName)
                            .const
                            .init(makeCall("component.content.getElementById", elementId).build())
                            .build())

                        .add(ifStatement(unaryExpression("!", identifier(elementName)), makeThrow(`missing element: @${elementId}`)))
                        .add(makeAssignment(`${elementName}.id`, elementOriginalId))
                        .add(makeCall("autorun", declareFunction()
                            .body(declareObjectDestruction(...externalReferences).const.init(identifier("data")).build())
                            .body(...tokenSet.map(token => makeCall(elementName + ".setAttribute", token.attributeName, makeTemplateLiteral(token.content)).build()))
                            .build()).build())
                        ;
                }
            }

            funcBody.add(ifStatement(identifier("host"), expressionStatement(callExpression(identifier("host.appendChild"), [identifier("component.content")]))))
                .add(returnStatement(identifier("component.content")));


            componentFile.componentStatements.add(
                declareFunction(template.id)
                    .param("data")
                    .param("host", null, undefined, "Element")
                    .body(funcBody)
                    .export
                    .build());


        }
    }
    return true;
}
