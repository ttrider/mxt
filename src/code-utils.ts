import { Element } from "domhandler";
import { getOuterHTML } from "DomUtils";
import { parseSync } from "@babel/core";
import * as t from "@babel/types";

interface IBuilder {
    build: t.Statement | t.Expression
}

function isBuilder(value: any): value is IBuilder {
    return (value && value.build);
}

export declare type StatementItem =
    t.Statement |
    t.Expression |


    { statements: t.Statement[] }
    ;

export function statementList(statements?: t.Statement[], ...node: Array<StatementItem>) {

    const items = (statements === undefined) ? [] : statements;

    const data =
    {
        statements: items,
        add
    };

    return add(...node);

    function add(...node: Array<StatementItem>) {
        for (let statement of node) {

            if (t.isStatement(statement)) {
                items.push(statement);
            }
            else if (t.isExpression(statement)) {
                items.push(t.expressionStatement(statement));
            }
            else {
                if (statement.statements) {
                    items.push(...statement.statements);
                }
            }

        }

        return data;
    }
}

export function declareFunction(functionName?: string) {

    const parameters: t.Identifier[] = [];
    const body = statementList();
    let isExport = false;

    const context: IBuilder & any = {
        get export() { isExport = true; return context },


        param: addParam,
        body: addStatement,

        get statement() { return buildStatement() }
    };

    return context;

    function buildStatement() {

        // identifiy optional parameters
        let canBeOptional = true;
        for (let i = parameters.length - 1; i >= 0; i--) {
            if (!canBeOptional) {
                parameters[i].optional = false;
            } else {
                if (!parameters[i].optional) {
                    canBeOptional = false;
                }
            }
        }

        if (functionName) {

            const func = t.functionDeclaration(t.identifier(functionName), parameters, t.blockStatement(body.statements));
            if (isExport) {
                return t.exportNamedDeclaration(func, []);
            }
            return func;

        }
        return t.arrowFunctionExpression(parameters, t.blockStatement(body.statements))
    }

    function addStatement(...node: Array<StatementItem>) {
        body.add(...node);
        return context;
    }


    function addParam(name: string, ...types: (null | string | undefined)[]) {

        const i = t.identifier(name);

        if (types.length === 0) {
            i.typeAnnotation = t.tsTypeAnnotation(t.tsAnyKeyword());
        } else if (types.length === 1) {
            i.typeAnnotation = t.tsTypeAnnotation(getType(types[0]));
        } else {
            i.typeAnnotation = t.tsTypeAnnotation(t.tsUnionType(types.map(getType)));
        }

        parameters.push(i);

        return context;

        function getType(tp: null | string | undefined) {

            if (tp === undefined) {
                i.optional = true;
                return (t.tsUndefinedKeyword());
            }

            if (tp === null) {
                return (t.tsNullKeyword());
            }



            return (t.tsTypeReference(t.identifier(tp)));
        }

    }




}

export function declareVar(name: string) {

    let kind: "const" | "let" | "var" = "const";
    let initExpression: t.Expression;

    const context: IBuilder & any = {
        get const() { kind = "const"; return context },
        get let() { kind = "let"; return context },
        get var() { kind = "var"; return context },

        init: addInitValue,

        get statement() { return buildStatement() }
    };

    return context;

    function addInitValue(value: any) {
        initExpression = getValueExpression(value);
        return context;
    }

    function buildStatement() {
        return t.variableDeclaration(kind, [
            t.variableDeclarator(
                t.identifier(name),
                initExpression)
        ]);
    }
}

export function declareObjectDestruction(...props: Array<{ name: string, as?: string }>) {

    let kind: "const" | "let" | "var" = "const";
    let initExpression: t.Expression;
    const properties: Array<{ name: string, as?: string }> = [...props];

    const context: IBuilder & any = {
        get const() { kind = "const"; return context },
        get let() { kind = "let"; return context },
        get var() { kind = "var"; return context },

        property: addProperty,

        init: addInitValue,

        get statement() { return buildStatement() }
    };

    return context;

    function addProperty(name: string, as?: string) {
        properties.push({ name, as });
        return context;
    }

    function addInitValue(value: any) {
        initExpression = getValueExpression(value);
        return context;
    }

    function buildStatement() {

        return t.variableDeclaration(kind, [
            t.variableDeclarator(
                t.assignmentPattern(
                    t.objectPattern(
                        properties.map(er => t.objectProperty(t.identifier(er.name), t.identifier(er.as ? er.as : er.name), undefined, !er.as))),
                    initExpression))
        ]);

    }
}

export function makeCall(name: string, ...params: any[]) {

    const body = statementList();
    const parameters = params.map(getValueExpression);

    const context: IBuilder & any = {
        param: addParam,

        get statement() { return buildStatement() }
    };

    return context;

    function addParam(...params: any[]) {
        parameters.push(...params.map(getValueExpression));
    }

    function buildStatement() {

        return t.callExpression(
            t.identifier(name),
            parameters);
    }

}
export function makeAssignment(target: string, value: any) {
    return t.assignmentExpression("=", t.identifier(target), getValueExpression(value));

}

export function makeTemplateLiteral(literal: string | Element[]) {

    if (typeof literal !== "string") {
        literal = getOuterHTML(literal, { decodeEntities: true, xmlMode: true });
    }
    return t.identifier("`" + literal + "`");
}

export function makeThrow(message: string) {

    return t.throwStatement(
        t.newExpression(t.identifier("Error"), [t.stringLiteral(message)])
    );
}


function parseStatements(code: string, wrapAsTemplate?: boolean) {

    if (wrapAsTemplate) {
        code = "`" + code + "`";
    }

    const results = parseSync(code, {
        babelrc: false,
        configFile: false
    });



    if (results && results.type === "File" && results.program) {
        return results.program.body;
    }

}


function getValueExpression(value?: any) {
    if (value === undefined) {
        return t.identifier("undefined");
    }
    else if (t.isExpression(value)) {
        return value;
    }
    return t.valueToNode(value);
}



