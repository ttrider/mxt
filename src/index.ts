import glob from "glob";
import util from "util";
import fs from "fs";
import path from "path";
import { parseDOM } from "htmlparser2";
import { Node, Element } from "domhandler";
import { ElementType } from "domelementtype";
import { Handler } from "htmlparser2/lib/Parser";

const globp = util.promisify(glob);
const readFile = util.promisify(fs.readFile);

import { parseSync, parse, traverse } from "@babel/core";
import { statement } from "@babel/template";
import { Expression, SpreadElement, Statement, Scopable, V8IntrinsicIdentifier, JSXNamespacedName, ArgumentPlaceholder, VariableDeclaration } from "@babel/types";





declare type PluginHandlers =
    "beforeLoad" |
    "afterLoad" |
    "beforePackage" |
    "package" |
    "afterPackage" |
    "done";

declare type PluginComponentHandlers =
    "beforeCodegen" |
    "codegen" |
    "afterCodegen";


interface PluginHandlerMap {
    "beforeTransform": Node;
    "transform": Node;
    "afterTransform": Node;

    "beforeTransformStyle": Element;
    "transformStyle": Element;
    "afterTransformStyle": Element;

    "beforeTransformLink": Element;
    "transformLink": Element;
    "afterTransformLink": Element;

    "beforeTransformScript": Element;
    "transformScript": Element;
    "afterTransformScript": Element;

    "beforeTransformTemplate": Element;
    "transformTemplate": Element;
    "afterTransformTemplate": Element;

    "beforeParseStyle": Element;
    "parseStyle": Element;
    "afterParseStyle": Element;

    "beforeParseScript": Element;
    "parseScript": Element;
    "afterParseScript": Element;

    "beforeParseTemplate": Element;
    "parseTemplate": Element;
    "afterParseTemplate": Element;
}

export interface ElementInfo {
    name: string,
    attributes: { [name: string]: string },
    content?: string,
    expressions?: string[];

}
export interface StyleElementInfo extends ElementInfo {
    rules: string[]
}

export interface TemplateInfo extends ElementInfo {
    id: string,
    elements: Element[],
    tokens: AttributeTokenInfo[];
}

export interface ExpressionInfo {
    content: string,
    hasTokens: boolean,
    externalReferences: string[]
}

export interface AttributeTokenInfo extends ExpressionInfo {
    elementId: string;
    attributeName: string;
}

export interface Options {
    input: string;

    plugins?: Plugin[];

    //es5 vs es6

}

interface Plugin {
    beforeLoad?: (context: HandlerContext) => boolean;
    afterLoad?: (context: HandlerContext) => boolean;


    beforeTransform?: (context: HandlerContext, componentFile: ComponentFileInfo, node: Node) => boolean;
    transform?: (context: HandlerContext, componentFile: ComponentFileInfo, node: Node) => boolean;
    afterTransform?: (context: HandlerContext, componentFile: ComponentFileInfo, node: Node) => boolean;

    beforeTransformStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeTransformLink?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformLink?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformLink?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeTransformScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeTransformTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeParseStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    parseStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterParseStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeParseScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    parseScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterParseScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeParseTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    parseTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterParseTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeCodegen?: (context: HandlerContext, componentFile: ComponentFileInfo) => boolean;
    codegen?: (context: HandlerContext, componentFile: ComponentFileInfo) => boolean;
    afterCodegen?: (context: HandlerContext, componentFile: ComponentFileInfo) => boolean;

    beforePackage?: (context: HandlerContext) => boolean;
    package?: (context: HandlerContext) => boolean;
    afterPackage?: (context: HandlerContext) => boolean;
    done?: (context: HandlerContext) => boolean;
}

export interface HandlerContext {

    options: Options,
    files: string[],
    componentFiles: ComponentFileInfo[],
    plugins: Plugin[],

    globalLinkElements: ElementInfo[],
    globalStyleElements: ElementInfo[],
    bodyElements: ElementInfo[]
}


export interface ComponentFileInfo {

    path: string,
    name: string,
    content: string,
    dom: Node[],
    links: ElementInfo[],
    globalStyles: StyleElementInfo[],

    templates: { [id: string]: TemplateInfo },
    errors: Error[]
}



async function mxt(options: Options) {

    options.plugins = (options.plugins || []);

    const files = await globp(options.input, { absolute: true });

    const context: HandlerContext = {
        options: options,
        files,
        plugins: (options.plugins || []),
        componentFiles: [],
        globalLinkElements: [],
        globalStyleElements: [],
        bodyElements: []
    };

    execute("beforeLoad", context);

    await loadFiles(context);

    execute("afterLoad", context);

    // transform stage
    for (const componentFile of context.componentFiles) {
        transformFile(context, componentFile);
    }
    // parse stage
    for (const componentFile of context.componentFiles) {
        parseFile(context, componentFile);
    }
    // codegen stage
    for (const componentFile of context.componentFiles) {
        codegenFile(context, componentFile);
    }

    execute("beforePackage", context);
    execute("package", context);
    execute("afterPackage", context);
    execute("done", context);
}
export default mxt;

async function loadFiles(context: HandlerContext) {

    for (const filePath of context.files) {

        const fileName = path.basename(filePath);
        const ext = path.extname(fileName);

        const fileBuffer = await readFile(filePath);
        if (fileBuffer) {

            const content = fileBuffer.toString();

            context.componentFiles.push({
                path: filePath,
                name: fileName.substr(0, fileName.length - ext.length),
                content: content,
                dom: parseDOM(content, { xmlMode: true, withStartIndices: true, withEndIndices: true }),
                links: [],
                globalStyles: [],
                templates: {},
                errors: []
            });
        }

    }
}

function transformFile(context: HandlerContext, componentFile: ComponentFileInfo) {

    for (const node of componentFile.dom) {

        executeComponentNode("beforeTransform", context, componentFile, node);
        executeComponentNode("transform", context, componentFile, node);

        if (node.type) {
            const nodeType = node.type.toLowerCase();

            switch (nodeType) {
                case "style":
                    executeComponentNode("beforeTransformStyle", context, componentFile, node as Element);
                    executeComponentNode("transformStyle", context, componentFile, node as Element);
                    executeComponentNode("afterTransformStyle", context, componentFile, node as Element);
                    break;
                case "link":
                    executeComponentNode("beforeTransformLink", context, componentFile, node as Element);
                    executeComponentNode("transformLink", context, componentFile, node as Element);
                    executeComponentNode("afterTransformLink", context, componentFile, node as Element);
                    break;
                case "script":
                    executeComponentNode("beforeTransformScript", context, componentFile, node as Element);
                    executeComponentNode("transformScript", context, componentFile, node as Element);
                    executeComponentNode("afterTransformScript", context, componentFile, node as Element);
                    break;
                case "template":
                    executeComponentNode("beforeTransformTemplate", context, componentFile, node as Element);
                    executeComponentNode("transformTemplate", context, componentFile, node as Element);
                    executeComponentNode("afterTransformTemplate", context, componentFile, node as Element);
                    break;
            }

        }

        executeComponentNode("afterTransform", context, componentFile, node);
    }
}

function parseFile(context: HandlerContext, componentFile: ComponentFileInfo) {

    for (const node of componentFile.dom) {

        if (node.type) {
            const nodeType = node.type.toLowerCase();

            switch (nodeType) {
                case "link":
                case "style":
                    executeComponentNode("beforeParseStyle", context, componentFile, node as Element);
                    executeComponentNode("parseStyle", context, componentFile, node as Element);
                    executeComponentNode("afterParseStyle", context, componentFile, node as Element);
                    break;
                case "script":
                    executeComponentNode("beforeParseScript", context, componentFile, node as Element);
                    executeComponentNode("parseScript", context, componentFile, node as Element);
                    executeComponentNode("afterParseScript", context, componentFile, node as Element);
                    break;
                case "template":
                    executeComponentNode("beforeParseTemplate", context, componentFile, node as Element);
                    executeComponentNode("parseTemplate", context, componentFile, node as Element);
                    executeComponentNode("afterParseTemplate", context, componentFile, node as Element);
                    break;
            }
        }
    }
}

function codegenFile(context: HandlerContext, componentFile: ComponentFileInfo) {
    executeComponent("beforeCodegen", context, componentFile);
    executeComponent("codegen", context, componentFile);
    executeComponent("afterCodegen", context, componentFile);
}



function execute(handlerName: PluginHandlers, context: HandlerContext) {
    for (const plugin of context.plugins) {
        const handler = plugin[handlerName];
        if (handler && handler(context)) {
            return true;
        }
    }
    return false;
}
function executeComponent(handlerName: PluginComponentHandlers, context: HandlerContext, componentFile: ComponentFileInfo) {

    for (const plugin of context.plugins) {
        const handler = plugin[handlerName];
        if (handler && handler(context, componentFile)) {
            return true;
        }
    }
    return false;
}
function executeComponentNode<T extends keyof PluginHandlerMap>(handlerName: T, context: HandlerContext, componentFile: ComponentFileInfo, element: PluginHandlerMap[T]) {

    for (const plugin of context.plugins) {
        const handler = plugin[handlerName];
        if (handler && handler(context, componentFile, element as any)) {
            return true;
        }
    }


    return false;
}

export function extractExpressions(content: string) {

    const expressions: string[] = [];

    const regex = /([^\\]|^)\${([^}]*)}/gm;

    let m;

    while ((m = regex.exec(content)) !== null) {
        // This is necessary to avoid infinite loops with zero-width matches
        if (m.index === regex.lastIndex) {
            regex.lastIndex++;
        }

        expressions.push(m[2]);
    }

    return expressions;
}

export function parseExpressions(content: string) {

    const results = parseSync("`" + content + "`", {
        babelrc: false,
        configFile: false
    });
    if (!results) return;

    const state: ExpressionInfo = {
        content,
        hasTokens: false,
        externalReferences: []
    };

    traverse(results, {
        TemplateLiteral(path) {
            if (path.node.expressions.length > 0) {
                state.hasTokens = true;
            }
        },
        Identifier(path, state: ExpressionInfo) {
            const name = path.node.name;
            if (!path.scope.hasBinding(name)) {
                state.externalReferences.push(name);
            }
        },
    }, undefined, state);

    return state;
}


const defaultHandlers: Handler[] = [

];


