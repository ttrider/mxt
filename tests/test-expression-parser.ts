import { parseExpressions } from "../src";


test("null expression", () => {

    const results = parseExpressions("abc");
    expect(results).toMatchSnapshot();
});

test("simple value expression", () => {

    const results = parseExpressions("${abc}");
    expect(results).toMatchSnapshot();
});

test("multiple value expressions", () => {

    const results = parseExpressions("abc${def}ghi${jkl}");
    expect(results).toMatchSnapshot();
});

test("expression with functions", () => {

    const results = parseExpressions("${function(a){return a*b*c;}(d)}");
    expect(results).toMatchSnapshot();
});