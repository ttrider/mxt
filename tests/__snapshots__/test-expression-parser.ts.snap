// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression with functions 1`] = `
Object {
  "content": "\${function(a){return a*b*c;}(d)}",
  "externalReferences": Array [
    "b",
    "c",
    "d",
  ],
  "hasTokens": true,
}
`;

exports[`multiple value expressions 1`] = `
Object {
  "content": "abc\${def}ghi\${jkl}",
  "externalReferences": Array [
    "def",
    "jkl",
  ],
  "hasTokens": true,
}
`;

exports[`null expression 1`] = `
Object {
  "content": "abc",
  "externalReferences": Array [],
  "hasTokens": false,
}
`;

exports[`simple value expression 1`] = `
Object {
  "content": "\${abc}",
  "externalReferences": Array [
    "abc",
  ],
  "hasTokens": true,
}
`;
